fix(expo): increase compileSdkVersion to 31 and add EAS instructions to doc

abumalick opened the pull request:
## The problem

As mentioned in #314 the current version of Notifee does not build when using expo managed app with the expo plugin as the default `compileSdkVersion` is 30 and Notifee needs 31. Also EAS is using JDK 8 by default and using compileSdkVersion 31 requires the build to be done with JDK 11.

## What is added in this PR
- Expo plugin: Increase compileSdkVersion to 31 in project gradle file if the initial version is 30
- Expo installation documentation: Add a note about the required JDK 11 and the `eas.json` configuration to use JDK 11.

## What's left:
- [x] Run plugins tests and update snapshot
- [x] make an example repository that tests the plugin: https://github.com/abumalick/expo-notifee-example

ref #314

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/invertase/notifee?pullRequest=371) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/invertase/notifee?pullRequest=371) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/invertase/notifee?pullRequest=371) it.</sub>

abumalick commented:
@mikehardy Here is the PR that I proposed in #314 
I need help to build the project through, when I try to install packages using `yarn` I get this error:

```
error An unexpected error occurred: "ENOENT: no such file or directory, lstat '/Users/abumalick/workspace/coded/notifee/packages/react-native/node_modules/find-up/node_modules/path-exists'".
info If you think this is a bug, please open a bug report with the information provided in "/Users/abumalick/workspace/coded/notifee/yarn-error.log".
info Visit https://yarnpkg.com/en/docs/cli/install for documentation about this command.
```

If I run `ls /Users/abumalick/workspace/coded/notifee/packages/react-native/node_modules/find-up/node_modules/`, I see that the folder is empty. Are you are aware about this issue? Do you know a work around?

mikehardy commented:
This error has been frustrating me for a long time and has resisted all of my attempts to fix it so far. Workaround is to rm -rf the node modules in root, in tests_react_native and in packages/react-native. It installs correctly first time then fails on future times. Really irritating sorry. Something to do with the monorepo setup

mikehardy commented on packages/react-native/plugin/__tests__/androidProjectBuild.test.ts:
```diff
@@ -1,10 +1,25 @@
 
 describe('[Android] project build.gradle test', () => {
-  it('applies changes to project build.grade', async () => {
+  it('applies maven repository to project build.grade', async () => {
```
```suggestion
  it('updates java compileSdkVersion to 31 in project build.gradle', async () => {
```

mikehardy commented on packages/react-native/plugin/__tests__/androidProjectBuild.test.ts:
```diff
@@ -1,10 +1,25 @@
+    console.log(projectGradle);
+  });
+
+  it('applies maven repository to project build.grade', async () => {
```
```suggestion
  it('adds custom maven repository to project build.gradle', async () => {
```

mikehardy reviewed (changes requested):
I think this looks great - there may be a test fixture to update? At this line I think https://github.com/invertase/notifee/blob/fa7f1f01546e6e1e6519d1f5b3533baf18161c1e/packages/react-native/plugin/__tests__/__snapshots__/androidProjectBuild.test.ts.snap#L10

And I left some simple suggestions for test names plus a thought on defaulting to a fixed EAS image. What do you think?

abumalick commented on docs/react-native/docs/installation.md:
```diff
@@ -78,3 +78,19 @@ Then, add `@notifee/react-native` to the list of plugins in your app's Expo conf
+      "developmentClient": true,
+      "distribution": "internal",
+      "android": {
+        "image": "latest"
```
Yes it makes sense 👍🏻

abumalick commented:
@mikehardy I don't understand how I should run the tests inside `packages/react-native/plugins`. `yarn tests_rn:test` does not run these tests ... Should we fix the jest configuration to run these tests as well?

mikehardy commented:
> @mikehardy I don't understand how I should run the tests inside `packages/react-native/plugins`. `yarn tests_rn:test` does not run these tests ... Should we fix the jest configuration to run these tests as well?

Oh! This expo plugin is pretty new, and I don't Expo personally so I probably just did not notice they weren't running! I will trust your judgement on how to get them to run, but anything that does end up running them should work. That's a general statement. A specific statement is that hooking them to tests_rn:test does sound reasonable. Good catch!

mikehardy reviewed (approved):
This looks great and I just want to marks it as kind of "preiminary approval" - looks like there is still a little bit of inspection on how to get the unit test for the expo plugin to actually run in CI, but the code looks 💯 to me and that's worth noting

abumalick commented:
Thank you for the follow up @mikehardy , it is much appreciated.

- I pushed the changes to the config to run the tests and fixed the test fixture. I added a simple test for the plugin to improve the coverage. (It is still not 100%, if you want it to reach 100% I can do it with some quick refactoring)
- I created an example app that uses this plugin code, it is a bit dirty, I had to copy the plugin code to the project as I don't want to publish an npm version of the package. You can read the steps that I followed to create the example repository in the readme: https://github.com/abumalick/expo-notifee-example . I built successfully an apk in EAS from this example repository: https://expo.dev/accounts/abumalick/projects/expo-notifee-example/builds/818f0c0a-acd3-4300-94ad-b3b5c72b18b2

mikehardy commented:
Fantastic! It's a still the weekend here and a holiday tomorrow but I'll queue this for review. Thanks for the quick turnarounds, we should be able to get this in and published quickly

mikehardy reviewed (approved):
I'd like to try your test which is the part that will take a while, but while running some other tests just now I had time to read through this again - nice clean integration of the expo plugin tests in the jest setup

abumalick commented:
Thank you Mike. Take your time for the review, no problem.

I saw that you created a patch file of a pull request on rn facebook sdk
repository. Can you share with me how you generated it? I was wondering how
we can do that with generated files that are not tracked by git.

On Sun, Apr 10, 2022 at 8:24 PM Mike Hardy ***@***.***> wrote:

> ***@***.**** approved this pull request.
>
> I'd like to try your test which is the part that will take a while, but
> while running some other tests just now I had time to read through this
> again - nice clean integration of the expo plugin tests in the jest setup
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/invertase/notifee/pull/371#pullrequestreview-937325746>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AB75GYJTWTYATSABNLTFBWLVEM2IPANCNFSM5S6VXINQ>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

mikehardy commented:
I've been meaning to add something like this https://github.com/invertase/react-native-firebase/blob/main/.github/workflows/create_test_patches.yml (works in a monorepo) or this https://github.com/invertase/react-native-google-mobile-ads/blob/main/.github/workflows/create_test_patches.yml (for single package repos) here but haven't gotten to it yet.

They show the basic structure

For me what I do is use whatever example app the repo has typically, install my desired version (via git reference or from local files) build it and test it to make sure it's right, then run patch-package in that directory after doing an `./gradlew clean` in the android sub-directory (to clean out android build files). That's what I did for react-native-fbsdk-next

mikehardy reviewed (approved):
This still looks good to me - I haven't had time to run the test and will not for a while, but I don't see a reason to delay it. Thanks a bunch for this!

abumalick commented:
Thank you very much!

On Thu, Apr 21, 2022, 9:23 PM Mike Hardy ***@***.***> wrote:

> Merged #371 <https://github.com/invertase/notifee/pull/371> into main.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/invertase/notifee/pull/371#event-6474139690>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AB75GYK5PMW54YATW4YBLF3VGHBLVANCNFSM5S6VXINQ>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>
